src/modules/multiblade/clustering/ReferenceDataTest.cpp

```diff
@@ -0,0 +1,349 @@
+#include <assert.h>
+#include <math.h>
+#include <stdlib.h>
+
+// #undef TRC_LEVEL
+// #define TRC_LEVEL TRC_L_DEB
+
+using namespace Multiblade;
+
+
+// Specific to MB16 and MB18 prototypes
+const uint16_t NumberOfWires{32};
+
+// this will conditinally include the large datasets
+// and create 'unit tests' for this.
+// #define INCLUDE_DS1
+// #define INCLUDE_DS1_FILTERED
+// #define INCLUDE_DS2
+
+#include <ReferenceDataTestData.h>
+
+class ReferenceDataTest : public TestBase {
+public:
+  const bool NoFiltering{false};
+  const bool DoFiltering{true};
+  const bool DoSorting{true};
+
+  struct {
+    uint32_t Readouts{0}; // total number of readouts
+    uint32_t NoCoincidence{0}; // clusters with only strips or wires
+    uint32_t MatchedClusters{0}; // clusters with both strips and wires
+    uint32_t MatchedEvents{0}; // EFU - FP common events
+    uint32_t TimerWraps{0}; // Count when CAEN timer resets
+    uint32_t EventsInvalidWireGap{0}; // Events with wire gaps
+    uint32_t EventsInvalidStripGap{0}; // Events with strip gaps
+    uint32_t EventsInvalidThresh{0}; // Events with failed (wire) threshold
+  } Stats;
+
+  // Zero out stat counters
+  void clearStats() { std::fill_n((char*)&Stats, sizeof(Stats), 0); }
+
+  // Create a list of unmatched but 'good' events
+  // std::vector<Event> EventsExtra;
+
+  // Detect large jumps in readout time
+  void FixJumpsAndSort(std::vector<struct MBHits> & vec, bool sort);
+
+  // Do the clustering and event calulations
+  void LoadAndProcessReadouts(std::vector<struct MBHits> & vec);
+
+  // Compare previously calculated events with reference interpretation
+  // print out some stats
+  void CountMatches(std::vector<struct MBEvents> & evts, bool DiscardThresh);
+
+  bool thresholdCheck(uint8_t DigIndex, uint16_t GlobalChannel, uint16_t AdcValue);
+
+protected:
+  // builder uses a 2us time boxing
+  EventBuilder builder{2010};
+};
+
+bool compareByTime(const struct MBHits & a, const struct MBHits & b) {
+  return a.Time < b.Time;
+}
+
+bool isEqual(float t, uint16_t x, uint16_t y, struct MBEvents & res) {
+  return (fabsf(t - res.time) < 0.0000009) // 900ns
+          and ((uint16_t)(std::round(res.x)) == x)
+          and ((uint16_t)(std::round(res.y)) == y);
+}
+
+
+// MB18 specific
+bool isWire(uint16_t channel) {
+  return channel < NumberOfWires;
+}
+
+// MB18 specific - wires and strips share channel space, but we need each
+// coordinate to start at 0
+uint16_t GetWireCoord(uint32_t GlobalChannel) {
+  assert(isWire(GlobalChannel));
+  return GlobalChannel;
+}
+
+uint16_t GetStripCoord(uint32_t GlobalChannel) {
+  assert(GlobalChannel >= NumberOfWires);
+  return GlobalChannel - NumberOfWires;
+}
+
+bool ReferenceDataTest::thresholdCheck(uint8_t DigIndex, uint16_t GlobalChannel, uint16_t AdcValue) {
+  if (not isWire(GlobalChannel)) {
+    uint16_t StripCh = GlobalChannel;
+    if (AdcValue < 1) {
+      XTRACE(CLUSTER, DEB, "Strip threshold of 0 not met for ch %u ", StripCh);
```

Comment:
Is this a descriptive message for the case? This could be produced when `AdcValue` is 0, as the threshold tested in the conditional is 1.

Reply:
nope, will fix